### DISCUSSION TITLE
Proposal: rename `tool.pdm.scripts` to `tool.pdm.tasks`

### DISCUSSION THREAD

**Discussion by GabDug** [pdm-project/pdm#2540] [General]
Hey :wave:
I've recently been working on a VS Code extension for PDM and I noticed something: I'm not a fan of the naming of `tool.pdm.scripts`.

Thus I open a conversation about renaming it to `tool.pdm.tasks`, feel free to tell me what you think, whether your agree or disagree!

# Proposal: Rename `tool.pdm.scripts` to `tool.pdm.tasks`

## Summary

PDM supports setting "scripts" in `pyproject.toml` file, which are available to run development tasks.
However, the name "scripts" can cause confusion with the `project.script`, used for packaging.

Furthermore, PDM "scripts" can be commands, shell scripts, python functions or a combination of them, which is more like a task than a script.

I propose to rename `tool.pdm.scripts` to `tool.pdm.tasks`, which is more consistent with the concept of task in other tools, such as VSCode tasks, or Poe the poet tasks.

### Pros

- "Task" has better semantics than "script", considering the possibilities of the "scripts" in PDM.
- "Task" is more consistent with the concept of task in other tools, such as VSCode tasks, or Poe the poet tasks.
- "Task" would avoid confusion with "project.scripts", which is a different concept: tasks are for development, scripts are for packaging.

### Cons

- "Task" is a more generic term, and it could be confused with other concepts, such as "task" in asyncio, or "task" in Celery.
- Users coming from npm could be confused, because npm scripts are called "scripts".
- Burden of migration for users and maintainers.

## Ecosystem impact

Tools reading directly from `pyproject.toml` `tool.pdm.scripts` that are not updated may stop working.

Apart from this [VS Code extension](https://marketplace.visualstudio.com/items?itemName=knowsuchagency.pdm-task-provider), I'm not aware of tools using it directly. 

## Retrocompatibility

This change could be made in a retrocompatible way, to make it easier on user.

PDM could support both naming for a little while, so users can migrate their projects and users with different versions of PDM on the same project won't have issues.

I can take the responsibility to update PDM and its documentation to support the change.

### Deprecation plan

- Support both "scripts" and "tasks" in `pyproject.toml` file for N versions.
  - If both "scripts" and "tasks" are defined, the "scripts" will be ignored and a warning will be shown.
  - Deprecation warning for projects still using "tool.pdm.scripts" in `pyproject.toml` file.
  - The warning will show on startup? On some commands invocation?
- Automatic remediation after N versions.
  - If "tool.pdm.scripts" is found in `pyproject.toml` file, it will be renamed to "tool.pdm.tasks" and a warning will be shown.
  - If "tool.pdm.scripts" is found in `pyproject.toml` file, and "tool.pdm.tasks" is also found, a warning (error?) will be shown.

**Comment by jvllmr** [comment]
In the JavaScript ecosystem they are also called scripts, aren't they? At least `pnpm` and `yarn` (I don't know for npm) also have a hook mechanic for built-in tasks implemented. I agree that the name tasks fits better, but when looking at only the package manager ecosystem, `scripts` appears to be the standard established by `package.json`.

**Reply by jvllmr** [reply]
Only just read that you also mentioned that in  cons, oops 😅. But my point still stands.

**Reply by GabDug** [reply]
Yep, the `npm` ecosystem uses `scripts`, but they are limited to commands (no `composite`), and to my knowledge there's no equivalent to `project.scripts`, so no name confusion for them.

To summarize my point:
- I believe composition of "scripts" really makes it "tasks", akin to Makefile tasks
- The naming collision with `project.scripts` make things confusing and harder to explain

**Reply by baggiponte** [reply]
I don't have an opinion on this, but an anectode. I never knew `project.scripts` was a thing, though I knew about `tool.pdm.scripts`, so I tried to use `project.scripts` as it was a way to specify "task aliases".
